wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsDataSource.swift

```diff
@@ -165,16 +165,15 @@ final class MessageDetailsDataSource: NSObject, ZMMessageObserver, ZMUserObserve
 
     private func setupReactions() {
         reactions = message.usersReaction.map { reaction, users in
-            let emoji = Emoji(value: reaction)
-            return MessageDetailsSectionDescription(
-                headerText: "\(emoji.value) \(emoji.name?.capitalizingFirstCharacterOnly ?? "") (\(users.count))",
-                items: MessageDetailsCellDescription.makeReactionCells(users)
-            )
-        }.filter {
-            !$0.items.isEmpty
+            return (Emoji(value: reaction), users)
+        }.filter { _ , count in
+            return !count.isEmpty
         }
-
-        self.reactions.sort(by: { $0.items.count > $1.items.count })
+        .sortedByCountThenName()
+        .map({ (emoji, users) in
+            MessageDetailsSectionDescription(headerText: "\(emoji.value) \(emoji.name?.capitalizingFirstCharacterOnly ?? "") (\(users.count))",
+                                             items: MessageDetailsCellDescription.makeReactionCells(users))
+        })
```

Comment:
This is a lot of sequence transformations which can be expensive if there are a lot of reactions. Here are some ideas to make it more efficient.

By using `lazy` we can merge two transformations (filter and map) into one:

```swift
func sortedReactions() -> [(Emoji, [UserType])] {
    return usersReaction.lazy
        .filter { _, users in !users.isEmpty }
        .map { key, users in (Emoji(value: key), users) }
        .sortedByCountThenName()
}
```

Then you can do additional maps after than, such as to make the section description:

```swift
message.sortedReactions().map { emoji, users in
    ...
}
```

You can also use `sortedReactions` in other places wherever you have the message.

Reply:
Done! Thank you for the suggestions @johnxnguyen!